clang/lib/Tooling/DependencyScanning/DependencyScanningTool.cpp

```diff
@@ -203,6 +206,216 @@ DependencyScanningTool::getDependencyTreeFromCompilerInvocation(
   return FS.createTreeFromNewAccesses(RemapPath);
 }
 
+namespace {
+class IncludeTreePPConsumer : public PPIncludeActionsConsumer {
+public:
+  explicit IncludeTreePPConsumer(cas::CASDB &DB) : DB(DB) {}
+
+  Expected<cas::IncludeTreeRoot> getIncludeTree();
+
+  struct FileWithCharacteristic {
+    const FileEntry *FE;
+    SrcMgr::CharacteristicKind Characteristic;
+  };
+
+private:
+  void enteredInclude(Preprocessor &PP, FileID FID) override;
+
+  void exitedInclude(Preprocessor &PP, FileID IncludedBy, FileID Include,
+                     SourceLocation ExitLoc) override;
+
+  void handleHasIncludeCheck(Preprocessor &PP, bool Result) override;
+
+  Expected<cas::ObjectRef> getObjectForFile(Preprocessor &PP, FileID FID);
+  Expected<cas::ObjectRef>
+  getObjectForFileNonCached(FileManager &FM, const SrcMgr::FileInfo &FI);
+  Expected<cas::ObjectRef> getObjectForBuffer(const SrcMgr::FileInfo &FI);
+
+  struct FilePPState {
+    cas::ObjectRef File;
+    SmallVector<std::pair<cas::ObjectRef, unsigned>, 6> Includes;
+    llvm::SmallBitVector HasIncludeChecks;
+  };
+
+  Expected<cas::IncludeTree> getCASTreeForFileIncludes(FilePPState &&PPState);
+
+  bool hasErrorOccurred() const { return ErrorToReport.has_value(); }
+
+  template <typename T> Optional<T> check(Expected<T> &&E) {
+    if (!E) {
+      ErrorToReport = E.takeError();
+      return None;
+    }
+    return *E;
+  }
+
+  cas::CASDB &DB;
+  Optional<cas::ObjectRef> PCHRef;
+  llvm::BitVector SeenIncludeFiles;
+  Optional<cas::ObjectRef> PredefinesBufferRef;
+  SmallVector<FilePPState> IncludeStack;
+  llvm::DenseMap<FileWithCharacteristic, Optional<cas::ObjectRef>>
```

Comment:
Why is the `CharacteristicKind` needed in the key? I thought identity was only based on the `FileEntry *` for includes.

Reply:
Hmm, do you think this is unnecessarily conservative? From the `SourceManager`'s perspective `FileEntry` is orthogonal to `CharacteristicKind` but there's no instance where the preprocessor would call `SourceManager::createFileID()` and pass the same `FileEntry` with different `CharacteristicKind`, is this right?

Reply:
> Hmm, do you think this is unnecessarily conservative?

Yeah, that's what I'm wondering. The `SourceManager` stores file `ContentCache` based only on `FileEntry` (it also takes an `IsSystem` bit, but that's not part of the lookup key, only the value) and similarly `HeaderSearch` stores `HeaderFileInfo` only based on `FileEntry`.  Is there a reason include-tree needs to be more specific?

Reply:
Note that `SourceManager` keeps the `CharacteristicKind` separate from the `ContentCache`, it pairs them together in `SrcMgr::FileInfo`. You can create a `FileID` using the same `FileEntry` and different `CharacteristicKind` and "include" that `FileID`, and the preprocessor can actually do that:
```
#include "t.h" // included as C_User
# 4 "t.c" 3 // this file considered system from now on.
#include "t.h" // included as C_System
```
Assuming `t.h` can be included multiple times this creates 2 included `FileID`s with the same `FileEntry` but different `CharacteristicKind`.

Reply:
Could we store the `CharacteristicKind` in the `IncludeTree` instead of the `IncludeFile`? That allows you to remove it from the DenseMap key and only store it in the IncludeStack where it is cheaper (improves hashing perf, makes the DenseMap keys half the size).

Reply:
Seems like a good idea, I'll look into it.

Reply:
> Could we store the CharacteristicKind in the IncludeTree instead of the IncludeFile? 

I've pushed changes, this works better, thanks! 🙇🏻‍♂️